### ISSUE TITLE
[Feature] We need to extract GitLab and Gitee raw client

### ISSUE THREAD

**Issue by frank-zsy** [hypertrons/hypertrons#281]
Now the GitHub data raw client is a separate npm module (https://www.npmjs.com/package/github-data-cat), as we want to build digital panel for open source projects on different platform, we need to extract the GitLab and Gitee data raw client which is now in Hypertrons code.

The following items should be considered:

- Naming: maybe we should all name this kind of projects with data-cat-xxx, like data-cat-gitlab
- Publish: when publish to npm, we may need a Hypertrons namespace, so the packages can be publish to like @hypertrons/data-cat-gitlab
- Requests management: for all requests in Hypertrons project, we may need an identical promise management package to handle them, for example, bluebird(https://www.npmjs.com/package/bluebird). So all the independent modules should use the same promise library to make sure we can use a global promise manager to handle the requests(just like thread pool manager in Java) while integrating modules.

**Comment by GoodMeowing** [comment]
That's to say, we should determine the interface about timeout retry? For example, all data-cat receive the same form params to customize timeout time and retry time. And I notice that data-cat-github handle this part independently. Does it mean we should refactor it after our interface is confirm?

**Comment by frank-zsy** [comment]
> That's to say, we should determine the interface about timeout retry? For example, all data-cat receive the same form params to customize timeout time and retry time. And I notice that data-cat-github handle this part independently. Does it mean we should refactor it after our interface is confirm?

Yes, but above that, we should have an uniform promise handler, like a thread pool to handle the request, the most important feature of the handler maybe concurrency management.

**Comment by GoodMeowing** [comment]
Identical promise lib doesn't seem to be very important. For example, Bluebird will limit the number of concurrency only in a single API calling:
```
Promise.map(
  Iterable<any>|Promise<Iterable>,
  function(any item, int index, int length) mapper,
  [Object {concurrency: int=Infinity} options]
 ) -> Promise
```
The map method accepts an option for controlling concurrency. However, This kind of limit will take effect only in the same function calling. When platform clients send requests, those requests obviously happend in different places.

Thus we may need develop a independent manager instead of simply using outer package. Maybe we can set the control parameters (current concurrent requests number, request limit, timeout time, etc.) as a global singleton. Then pass it to all clients and make sure they can get latest infomation.

Even if we have to send all request in the management center, we still don't need to use same promise package. In this situation, all clients will not truely send requests, but will add tasks into a global request queue. Then manager will handle the queue.

So, the feature is something about the architecture design of sharing request management between different packages. Maybe I should search solutions from some distributed systems and discuss them in the meeting.

**Comment by liwen-tj** [comment]
Here is [gitlab-data-fetcher](https://github.com/liwen-tj/gitlab-data-fetcher). Maybe I still need to refine gitlab-graphql-client.

**Comment by frank-zsy** [comment]
@GoodMeowing if bluebird does not fit the requirement, you can look into other packages, like [`p-queue`](https://www.npmjs.com/package/p-queue) or [`bull`](https://www.npmjs.com/package/bull)

**Comment by GoodMeowing** [comment]
> @GoodMeowing if bluebird does not fit the requirement, you can look into other packages, like [`p-queue`](https://www.npmjs.com/package/p-queue) or [`bull`](https://www.npmjs.com/package/bull)

ok

**Comment by GoodMeowing** [comment]
[`p-queue`](https://www.npmjs.com/package/p-queue) can be the global request handler (mainly on concurrency), [`p-retry`](https://www.npmjs.com/package/p-retry) can control retry when encounter timeout or other error. 

We don't need to implement those functions manually, just add some customized codes.
1. Combine above packages into a global request handler.
2. Refator all client, make sure they send specific requests via the global handler. This handler is a option when a client is instantiated. Thus clients can work dependently or together.

**Comment by frank-zsy** [comment]
Yes, this is very good, you can make Gitee client to fit these kind of config param first and we can modify GitHub and GitLab client later.
